### PR TITLE
Bind API to convert from/to bytes to/from Variant

### PR DESCRIPTION
**Pull request by zeenix**
Bind Variant::new_from_bytes & Variant::get_data_as_bytes().

Related: #113.

### CODE REVIEW

File: src/variant.rs [hunk 1]
```diff
@@ -141,6 +142,19 @@ impl Variant {
             }
         }
     }
+
+    /// Constructs a new serialised-mode GVariant instance.
+    pub fn new_from_bytes<T: StaticVariantType>(bytes: &Bytes, trusted: bool) -> Self {
```
Comment (sdroege) [inline src/variant.rs]: If you pass random data and `trusted=true`, will it crash? If so this needs to be an `unsafe fn` (you probably want a safe variant with `trusted=false` and an unsafe with `trusted=true` then).

**Comment by sdroege** [comment]
Maybe also add the `to_bytes()` binding while you're at it? Then we keep it symmetric :)

File: src/variant.rs (hunk 1 above)
Comment (zeenix) [reply src/variant.rs]: I'll check.

**Comment by zeenix** [comment]
> Maybe also add the `to_bytes()` binding while you're at it? Then we keep it symmetric :)

Yeah, i intend to. I just wanted to get early feedback for learning purposes. :)

File: src/variant.rs (hunk 1 above)
Comment (zeenix) [reply src/variant.rs]: i tried a few things (mainly tried to give it less bytes and some curropted bytes) with `trusted=true` but i failed to make it crash. Also looking briefly through the code and docs, it seems unlikely that it skips critical checks when decoding the bytes so I think it's good like this.

Comment (sdroege) [reply src/variant.rs]: What is it actually doing then? I feel like this should get an issue reported for improving the docs against GLib :)

**Comment by zeenix** [comment]
> Maybe also add the `to_bytes()` binding while you're at it? Then we keep it symmetric :)

Done. I've not tested that one yet. I'll report it here once done.

**Comment by sdroege** [comment]
> Done. I've not tested that one yet. I'll report it here once done.

Looks good to me as long as that can never return NULL :)

File: src/variant.rs (hunk 1 above)
Comment (zeenix) [reply src/variant.rs]: Actually ebassi says it's UB so I'll go with your approach.

**Comment by zeenix** [comment]
> Looks good to me as long as that can never return NULL :)

Actually it can. :(

> If value is a fixed-sized value that was deserialised from a corrupted serialised container then NULL may be returned. In this case, the proper thing to do is typically to use the appropriate number of nul bytes in place of value . If value is not fixed-sized then NULL is never returned.

This is from docs of `g_variant_get_data` but `g_variant_get_data_as_bytes` docs say the semantics are identical.

**Comment by sdroege** [comment]
> Actually it can. :(

`Option` it is then :)

**Comment by zeenix** [comment]
All done.

**Comment by EPashkin** [comment]
From https://github.com/GNOME/glib/blob/mainline/glib/gvariant-core.c#L938 `g_variant_get_data_as_bytes` always return Bytes,
even in case NULL from `g_variant_get_data`
it doing something strange, but returns Bytes anyway.

**Comment by EPashkin** [comment]
@zeenix Sorry, I was wrong, `g_bytes_new_from_bytes` return NULL if it get wrong parameters, so `Option` is needed.
Thanks, 👍

**Comment by sdroege** [comment]
> @zeenix Sorry, I was wrong, `g_bytes_new_from_bytes` return NULL if it get wrong parameters, so `Option` is needed.

When?

**Comment by EPashkin** [comment]
If bytes_data==NULL, then `data - bytes_data=>data` so `g_bytes_new_from_bytes` get pointer as `offset` parameter and it not pass check in https://github.com/GNOME/glib/blob/mainline/glib/gbytes.c#L227

**Comment by sdroege** [comment]
That's a `g_return_val_if_fail()`, those are basically like panics. Nothing we need to take into account (other than trying to not make it possible to call C API in ways that end up in there).

**Comment by zeenix** [comment]
I removed the `Option`.

**Comment by sdroege** [comment]
Nice

>   process didn't exit successfully: `/home/travis/build/gtk-rs/glib/target/debug/deps/glib-b0a38ca36f7dfe57` (signal: 11, SIGSEGV: invalid memory reference)

On all nightly builds :see_no_evil:

**Comment by sdroege** [comment]
I can confirm, it's the `subclass::object::test::test_create` test that explodes. Unrelated to this PR though.

**Comment by sdroege** [comment]
Regression caused by https://github.com/rust-lang/rust/commit/fc23a81831d5b41510d3261c20c34dd8d32f0f31, hm!

**Comment by sdroege** [comment]
https://github.com/rust-lang/rust/issues/69102

**Comment by GuillaumeGomez** [comment]
Daily breaking changes time is back!

**Comment by sdroege** [comment]
Don't be so dramatic :) It's just a bug, and might even be one on our side that just happened to not trigger for a long long time.

**Comment by GuillaumeGomez** [comment]
My bad... I have memories from the past where every day was a whole new language haha.

**Pull request merged**